src/aliased_buffer.h

```diff
@@ -0,0 +1,191 @@
+
+#ifndef SRC_ALIASED_BUFFER_H_
+#define SRC_ALIASED_BUFFER_H_
+
+#include "v8.h"
+#include "util.h"
+#include "util-inl.h"
+
+namespace node {
+
+/**
+ * This class encapsulates the technique of having a native buffer mapped to
+ * a JS object. Writes to the native buffer can happen efficiently without
+ * going through JS, and the data is then available to user's via the exposed
+ * JS object.
+ *
+ * While this technique is computationaly efficient, it is effectively a
+ * write to JS program state w/out going through the standard
+ * (monitored) API. Thus any VM capabilities to detect the modification are
+ * circumvented.
+ *
+ * The encapsulation herein provides a placeholder where such writes can be
+ * observed. Any notification APIs will be left as a future exercise.
+ */
+template <class NativeT, class V8T>
+class AliasedBuffer {
+ public:
+  AliasedBuffer(v8::Isolate* isolate, const size_t count) :
+      isolate_(isolate),
+      count_(count),
+      byte_offset_(0),
+      free_buffer_(true) {
+    const v8::HandleScope handle_scope(isolate_);
+
+    const size_t sizeInBytes = sizeof(NativeT) * count;
+
+    // allocate native buffer
+    buffer_ = Calloc<NativeT>(count);
+
+    // allocate v8 ArrayBuffer
+    v8::Local<v8::ArrayBuffer> ab = v8::ArrayBuffer::New(
+        isolate_, buffer_, sizeInBytes);
+
+    // allocate v8 TypedArray
+    v8::Local<V8T> js_array = V8T::New(ab, byte_offset_, count);
+    js_array_ = v8::Global<V8T>(isolate, js_array);
+  }
+
+  /**
+   * Create an AliasedBuffer over a sub-region of another aliased buffer.
+   * The two will share a v8::ArrayBuffer instance &
+   * a native buffer, but will each read/write to different sections of the
+   * native buffer.
+   *
+   *  Note that byte_offset must by aligned by sizeof(NativeT).
+   */
+  AliasedBuffer(
+      v8::Isolate* isolate,
+      const size_t byte_offset,
+      const size_t count,
+      const AliasedBuffer<uint8_t,
+      v8::Uint8Array>& backing_buffer) :
+      isolate_(isolate),
+      count_(count),
+      byte_offset_(byte_offset),
+      free_buffer_(false) {
+    const v8::HandleScope handle_scope(isolate_);
+    // validate that the byte_offset is aligned with sizeof(NativeT)
+    CHECK_EQ(byte_offset & (sizeof(NativeT) - 1), 0);
```

Comment:
This is missing a range check to make sure the byte length of this instance fits within the `backing_buffer`. e.g.:
```cc
CHECK_LE(sizeof(NativeT) * count, backing_buffer.GetArrayBuffer()->ByteLength() - byte_offset);
```

Reply:
Fwiw, both that check and the existing check are also verified within V8

Reply:
@addaleax Possibly in a debug build, but I just created a module that creates a `Float64Array` and allowed me to have a count larger than the byte length of the array buffer.

Reply:
added check